tracing/src/main/java/com/palantir/tracing/TraceLocal.java

```diff
@@ -0,0 +1,152 @@
+/*
+ * (c) Copyright 2023 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.tracing;
+
+import com.palantir.logsafe.Preconditions;
+import com.palantir.logsafe.exceptions.SafeIllegalArgumentException;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+/**
+ * Provides trace local variables. Each trace gets its own independent copy of the variable.
+ *
+ * A trace local is either set (has a value) or unset (has no value).
+ *
+ * Outside of a trace (i.e. when {@link Tracer#hasTraceId()} is false) then the trace local is always unset.
+ */
+public final class TraceLocal<T> {
+
+    @Nullable
+    private final Function<? super TraceLocal<?>, T> initialValue;
+
+    private TraceLocal(@Nullable Supplier<T> initialValue) {
+        if (initialValue == null) {
+            this.initialValue = null;
+        } else {
+            // eagerly transform supplier to avoid allocation per invocation
+            // (computeIfAbsent takes a Function)
+            this.initialValue = _ignored ->
+                    Preconditions.checkNotNull(initialValue.get(), "TraceLocal initial value must not be null");
```

Comment:
Can you confirm my understanding here -- if the supplier/function return null, we'll throw in `TraceLocal::get` not here in constructor? I think we could allow the supplier to return `null` and the value will just not be added to the `traceLocals` map ([`ConcurrentHashMap` allows this](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/concurrent/ConcurrentHashMap.html#computeIfAbsent(K,java.util.function.Function))).

```suggestion
            this.initialValue = _ignored -> initialValue.get();
```

Reply:
yeah, throws in get() currently. Don't have strong feeling about what behaviour is best

Reply:
My goal with this suggestion is to ensure folks aren't using a supplier which returns null instead of using the factory with no supplier, since that one is more efficient in some cases (`get()` on an empty trace doesn't initialize a new map). I think the behavior is more consistent for users this way as well, but I think we should add a checkNotNull in the `withInitialValue` static factory to throw on ctor when a null supplier is provided.

It's easier to remove the constraint in a future commit than to add a constraint, so my preference was to start as strict as possible.

Reply:
will add the checknotnull in withInitialValue

Reply:
done

Reply:
I think this will need an approval from you, @schlosna, since some of my suggestions have been merged. I'm going to approve for now given the null check is currently set in the direction of least regret, but I'd like to confirm that you're happy with the current approach -- happy to change/revisit if you have a strong opinion.